### DISCUSSION TITLE
Updating build to support ARM/aarch64 with box86/box64

### DISCUSSION THREAD

**Discussion by lafrad** [mbround18/valheim-docker#733] [Ideas]
I'm interested in tweaking the docker container build to be able to run/build the container for native aarch64.
This would then build box86/box64 (when necessary), and run SteamCMD/Valheim on these compatibility layers.
The intent for myself is to run on ARM servers (Oracle/AWS clouds) but it could also support Pi (I do not have a Pi to test, however).

From what I can tell...  everything in the repo *should* be able to be built natively, (cargo/rust toolchain should be good?).  I don't *see* any binaries pulled down (other than valheim), but I may be missing something.

I'm pretty green when it comes to docker, and new to Rust.

A naive `docker compose up` within a local clone of the repo unsurprisingly spits out a build error which appears to be just cross platform issues:

```
 => ERROR [odin builder 7/7] RUN /usr/local/cargo/bin/cargo make -p production release                                                                                                        0.3s
------
 > [odin builder 7/7] RUN /usr/local/cargo/bin/cargo make -p production release:
0.248 /usr/local/cargo/bin/cargo-make: 1: ELF: not found
0.248 /usr/local/cargo/bin/cargo-make: 2: @%$�k: not found
0.249 /usr/local/cargo/bin/cargo-make: 3: �k: not found
0.249 /usr/local/cargo/bin/cargo-make: 4: p: not found
0.249 /usr/local/cargo/bin/cargo-make: 5: p: not found
0.249 /usr/local/cargo/bin/cargo-make: 6: p: not found
0.249 /usr/local/cargo/bin/cargo-make: 7: Syntax error: "(" unexpected
------
failed to solve: process "/bin/sh -c /usr/local/cargo/bin/cargo make -p production release" did not complete successfully: exit code: 2
```
(This is running/building on an `m7g.large` EC2 instance... ARM/Graviton aarch64, running ubuntu 22.04)
`Linux ip-... 6.2.0-1012-aws #12~22.04.1-Ubuntu SMP Thu Sep  7 16:00:15 UTC 2023 aarch64 aarch64 aarch64 GNU/Linux`

I'll start to unwind things a bit so things I can see pulled in can be compiled locally (instead of downloading binary builds)... but any tips on how to setup the `Dockerfile.<...>` and/or build settings to further this journey would be appreciated!

**Comment by lafrad** [comment]
Updating the `Dockerfile.odin` file to pull cargo-cheft and build (instead of download?) does cause more building, but I run into the same error:


```
diff --git a/Dockerfile.odin b/Dockerfile.odin
index d7c5161..f15e22a 100644
--- a/Dockerfile.odin
+++ b/Dockerfile.odin
@@ -1,10 +1,16 @@
 ARG DEBIAN_VERSION=12
 ARG RUST_VERSION=1.72

+FROM rust:1 AS chef
+# We only pay the installation cost once,
+# it will be cached from the second build onwards
+RUN cargo install cargo-chef
+WORKDIR app
+
 # ------------------ #
 # -- Odin Planner -- #
 # ------------------ #
-FROM lukemathwalker/cargo-chef:latest-rust-${RUST_VERSION} as planner
+FROM chef as planner
 WORKDIR /data/odin
 COPY . .
 RUN cargo chef prepare --recipe-path recipe.json
@@ -12,7 +18,7 @@ RUN cargo chef prepare --recipe-path recipe.json
 # ------------------ #
 # -- Odin Cacher  -- #
 # ------------------ #
-FROM lukemathwalker/cargo-chef:latest-rust-${RUST_VERSION} as cacher
+FROM chef as cacher
 WORKDIR /data/odin
 COPY --from=planner /data/odin/recipe.json recipe.json
 RUN cargo chef cook --release --recipe-path recipe.json
diff --git a/Dockerfile.valheim b/Dockerfile.valheim
index 9e95348..591c20f 100644
--- a/Dockerfile.valheim
+++ b/Dockerfile.valheim
```

```
 => ERROR [odin builder 7/7] RUN /usr/local/cargo/bin/cargo make -p production release                                                                                           0.3s
------
 > [odin builder 7/7] RUN /usr/local/cargo/bin/cargo make -p production release:
0.251 /usr/local/cargo/bin/cargo-make: 1: ELF: not found
0.252 /usr/local/cargo/bin/cargo-make: 2: @%$�k: not found
0.252 /usr/local/cargo/bin/cargo-make: 3: �k: not found
0.252 /usr/local/cargo/bin/cargo-make: 4: p: not found
0.252 /usr/local/cargo/bin/cargo-make: 5: p: not found
0.252 /usr/local/cargo/bin/cargo-make: 6: p: not found
0.252 /usr/local/cargo/bin/cargo-make: 7: Syntax error: "(" unexpected
------
failed to solve: process "/bin/sh -c /usr/local/cargo/bin/cargo make -p production release" did not complete successfully: exit code: 2
```

**Reply by lafrad** [reply]
This actually looks to be an ELF issue, not specifically relate to architecture....  So this would be... a toolchain thing?

**Comment by mbround18** [comment]
This will not work :/ you will have to have the Valheim team build a arm based executable before this project can support this.

**Reply by lafrad** [reply]
Bummer!

I was inspired for this by some of the docker images out in the community that are targeted at Pi: https://hub.docker.com/r/arokan/raspiheim  

I just have preferred this docker image as I'm using it elsewhere on x86 hardware.
